onnxruntime/core/optimizer/insert_cast_transformer.cc

```diff
@@ -8,12 +8,17 @@
 using namespace ONNX_NAMESPACE;
 using namespace ::onnxruntime::common;
 namespace onnxruntime {
+static bool IsMLFloat16Tensor(const NodeArg& node_arg) {
+  // Type() will return nullptr if node_arg.Exists() is true so don't need an additional check for that
+  return node_arg.Type() != nullptr &&
+         DataTypeImpl::TypeFromProto(*node_arg.TypeAsProto()) == DataTypeImpl::GetTensorType<MLFloat16>();
+}
+
 bool InsertCastTransformer::NeedInsertCast(const onnxruntime::Node* node, const onnxruntime::NodeArg* input) const {
-  // If the node's input is float16 and currently the node is not assigned to any XP.
-  // we need insert a cast to float, and put the node on CPU for default behavior.
+  // If the node's input is float16 and currently the node is not assigned to any EP
+  // we need to insert a cast to float, and put the node on CPU for default behavior.
   // TODO: a better check is to check does the CPU kernel with float exist or not.
-  return input->Type() != nullptr &&
-         DataTypeImpl::TypeFromProto(*input->TypeAsProto()) == DataTypeImpl::GetTensorType<MLFloat16>() &&
+  return IsMLFloat16Tensor(*input) &&
```

Comment:
Control flow nodes can have other EPs. We wrap them specially in the CUDA EP to avoid the data being copied to/from CPU across the node (logic obviously executes on CPU but the kernel registration is for the CUDA EP to avoid the data copy where possible)

Added check and comment in NeedInsertCast